src/Components/Tree/Component.purs

```diff
@@ -0,0 +1,171 @@
+          else
+            H.raise (ItemRemoved item itemPath)
+        pure a
+
+      SetItems items a -> do
+        H.modify_ _ { items = items, initial = items }
+        pure a
+
+      ToggleChildren indexPath a -> do
+        let pathLens = pathToLens indexPath _expanded
+        traverse_ (\l -> H.modify $ over l not) pathLens
+        pure a
+
+      SetSelections itemPaths a -> do
+        { items, initial } <- H.get
+        let paths = flip itemPathToIndexPath items <$> itemPaths
+            updates = (\p r -> r { items = expandPath p r.items }) <$> paths
+            updater = A.foldl (>>>) (_ { items = initial }) updates
+        H.modify_ updater
+        pure a
+
+    render :: State item -> H.ComponentHTML (Query item)
+    render { items, renderItem } =
+      HH.div_ $ A.concat $ A.mapWithIndex (renderRow 0 [] []) items
+      where
+        renderRow depth indexPath itemPath ix (Node { selected, expanded, children, value }) =
+          [ HH.div
+            [ css $ "border-b py-2 " <> ("pl-" <> (show (depth * 10))) ]
+            [ renderCarat children expanded (A.cons ix indexPath)
+            , HH.div
+              [ css "inline-flex" ]
+              [ Checkbox.checkbox_
+                [ HE.onChecked $ HE.input $ ToggleItem value itemPath (A.cons ix indexPath)
+                , HP.checked selected
+                ]
+                [ HH.fromPlainHTML $ renderItem value ]
+              ]
+            ]
+          ] <>
+          [ HH.div
+            [ css $ if expanded then "" else "hidden" ]
+            ( A.concat
+              $ A.mapWithIndex
+                (renderRow (depth + 1) (A.cons ix indexPath) (A.snoc itemPath value))
+                children )
+          ]
+
+        renderCarat children expanded path =
+          carat
+            [ HE.onClick $ HE.input_ (ToggleChildren path)
+            , css $ "mr-3 text-xl align-text-bottom cursor-pointer " <> visible
+            ]
+          where
+            carat = if expanded then Icon.caratDown else Icon.caratRight
+            visible = if A.length children > 0 then "visible" else "invisible"
+
+-----
+-- Helper functions for expanding paths, toggling checkboxes, etc.
+
+_items :: ∀ item. Lens' (State item) (Array (Node item))
+_items = prop (SProxy :: SProxy "items")
+
+pathToLens
+  :: ∀ p a
+   . Wander p
+  => IndexPath
+  -> Optic' p (Node a) Boolean
+  -> Maybe (Optic' p (State a) Boolean)
+pathToLens path lastProp = (<<<) _items <$> pathToLens'
+  where
+    pathToLens' :: Maybe (Optic' p (Array (Node a)) Boolean)
+    pathToLens' = A.foldl foldLens <$> (last <$> A.head path) <*> A.tail path
+
+    foldLens :: Optic' p (Array (Node a)) Boolean -> Int -> Optic' p (Array (Node a)) Boolean
+    foldLens l ix = Lens.ix ix <<< _children <<< l
+
+    last :: Int -> Optic' p (Array (Node a)) Boolean
+    last ix = Lens.ix ix <<< lastProp
+
+expandPath :: ∀ a. IndexPath -> Array (Node a) -> Array (Node a)
```

Comment:
❔ There was a TODO on this about changing the implementation? Is the TODO still relevant or have we decided it's not relevant anymore? If it's still wanted, can we add it back until we address it?

Reply:
Oh yeah, good catch!